Halon combustion releases resin foam and pluoxium. Halon is formed by electrolysing BZ, and the reaction is much quicker.

Pickle-Coding opened the pull request:
## About The Pull Request
Original PRs: #69196 and #74472

Applies both of those PRs to this PR.
Doesn't apply the firefighter backpack halon requirement changes since absolutely no one liked that.

Halon combustion will combust into 2.5 moles of pluoxium per mole of halon consumed, and release atmos resin. The resin will not only help choke out fires, but it will also help seal rooms from fires, or seal fires from other rooms. Halon formation now requires electrolysing BZ instead of CO2 and N2O. The electrolysis is much faster. 2 moles of halon and 0.2 moles of oxygen is released with 91.2321kJ of thermal energy per mole of BZ electrolysed. Halon combustion starts at 343.15K instead of 373.15K.

Resin foam no longer consumes halon and pluoxium. The resin foam released from halon combustion is a special type that lasts longer to ensure the halon can spread and not block itself when combusting.
## Why It's Good For The Game
I have only seen one, maybe two rounds ever where someone created halon, and 0 rounds where it was actually used. The main reason why it (at least I think) isn't used is because it takes a rather large amount of effort and time for something that is only going to be used in a fraction of rounds. Preemptively spending that much effort every round, where most of the time it doesn't get used, will get boring and repetitive for pretty much anyone very quickly. Creating it as a response to a fire is going to take too long because you need to build a special cooling chamber and a contraption to ensure low pressure, and then wait for it to react slowly.

This PR tries to resolve this by changing the requirement to BZ, making the electrolysis much quicker and scales with temperature instead of inversely with pressure, and making it combust into pluoxium. BZ is a gas that takes a little effort to make, but is probably the most frequently made gas and is likely already available when there is an emergency. With the BZ requirement combined with electrolysis speed change, creating halon is just a matter of dumping BZ into an electrolyser, turning the electrolyser on then extracting the halon, which can be reasonably done as a response to a plasmaflood or whatever.

Halon combusting into pluoxium also makes halon not a dead end on the tree of gases. By letting halon combust into pluoxium, atmospheric technicians may persue halon to create pluoxium, which should let it feel less of a waste of time if someone decides to make it and there's no emergency that calls for the use of it. Halon combustion releasing resin foam conveys a visual effect that it is strong at choking fires. Reducing the temperature allows it to react earlier to a spreading fire, so it can block it before it spreads.
## Changelog
:cl:
balance: Halon combustion releases 2.5 moles of pluoxium per mole of halon consumed.
balance: Halon combustion no longer releases carbon dioxide.
balance: Halon combustion releases resin foam.
balance: Halon formation requirements changed to require electrolysing BZ instead of N2O and CO2.
balance: Halon formation changed to no longer slow down at higher pressures. It instead accelerates at higher temperatures. It is much faster overall.
balance: Halon formation releases 91.2321kJ of thermal energy per mole of halon consumed.
/:cl:

Pickle-Coding committed:
Halon formation change.

Changes halon formation to only require BZ. Pressure is redundant, removes pressure var. Asserts bz, oxygen, and halon, as they are the relevant gases. Reaction efficiency is temperature based. Consumes BZ, produces oxygen and halon. Changes the factor list to account for these changes.

Pickle-Coding committed:
Halon combustion produces pluoxium.

Halon combustion produces pluoxium instead of CO2. 2.5 moles of pluoxium per halon consumed. Asserts pluoxium instead of CO2.

Pickle-Coding committed:
Halon formation creates more energy.

The energy created from halon formation is the energy needed to set the reaction's pivot temperature to T0C (273.15K).

Pickle-Coding committed:
Scale with working power.

Scales the reaction rate with working_power.

Pickle-Coding committed:
Whoops.

Changes the energy release so the pivot is actually 273.15K.

Pickle-Coding committed:
Merge branch 'master' of https://github.com/Pickle-Coding/tgstation into halon_formation

Pickle-Coding committed:
Halon combustion resin effect.

Applies the changes to halon combustion changes and some of the resin foam changes from the Pickle-Coding:yo branch (up to commit 369c87cc11d2fdada74de5e4ff017f929c87f04d).

Also adds pluoxium to the ignored_gases list for /obj/structure/foamedmetal/resin since that's the product of halon combustion now.

PaznerIV commented:
It doesn't get made at least in my opinion because it's not really worth it. A good majority of rounds don't have fire and with how slow air moves I doubt it'll put it out quickly. Not to mention that fire extinguisher are a thing and they are easier to get that halon and are able to put out a fire at a moment notice

Jackal-boop commented:
Make fire extinguishers less effective <3

Pickle-Coding commented:
> Make fire extinguishers less effective <3

That would be for a separate PR (not anytime soon due to the scale of how fucked up it is) once I make reagents obey the laws of thermodynamics better.

iansdoor commented:
Praise Halong.

Pickle-Coding committed:
Factor fix.

Fixes the energy factor for halon formation not using the HALON_FORMATION_ENERGY define.

Pickle-Coding committed:
Remove pointless space.

Pickle-Coding committed:
Okay???

Not sure why it pissed about it being a non-constant expression when the define is a number. I'll just put the number in and leave it at that for now.

LemonInTheDark commented on code/__DEFINES/reactions.dm:
```diff
@@ -176,15 +176,17 @@
-/// The amount of energy 4.25 moles of halon forming from tritium and BZ releases.
-#define HALON_FORMATION_ENERGY 300
+/// Energy released per mole of BZ consumed during halon formation.
+#define HALON_FORMATION_ENERGY 91232.1
```
why so specific?

Pickle-Coding replied:
It's been like a year since I wrote this, so I don't remember. The most likely case is that the heat capacity of the reaction rises, so I chose an energy release that would converge the temperature to a specific number.

LemonInTheDark replied:
oh you're dojn your conservation of energy stuff again, thought that might be the case

LemonInTheDark replied:
is it actually reasonable?

Pickle-Coding replied:
Not a conservation of energy thing, more of a balancing thing so that the reaction stabilises itself to get to a temperature where it reacts at a reasonable speed, but still has room for optimising if an atmos tech desires.

LemonInTheDark commented on code/__DEFINES/reactions.dm:
```diff
@@ -176,15 +176,17 @@
+/// The minimum temperature required for halon to combust.
+#define HALON_COMBUSTION_MIN_TEMPERATURE (T0C + 70)
+/// The number of moles combusted required to release resin.
+#define HALON_COMBUSTION_MINIMUM_RESIN_MOLES 0.09
```
similar

Pickle-Coding replied:
Wanted to give some room for halon to react before plasma reaches combustion temperature, so it's better at trapping a spreading plasma fire.

Halon combustion consumes halon slowly, so I chose a rather small value for the requirements, without having like 1 fempto mole of halon release resin.

LemonInTheDark replied:
I mean like why is it not 0.1 why is it 0.09

Pickle-Coding replied:
Oh I remember. 0.1 moles would be too much for the combustion rate near the minimum temperature.

Pickle-Coding replied:
I composed the define with other defines so the reasoning should be more obvious now.

LemonInTheDark commented on code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm:
```diff
@@ -450,6 +448,15 @@
+/// A variant of resin foam that is created from halon combustion. It does not dissolve in heat to allow the gas to spread before foaming.
+/obj/effect/particle_effect/fluid/foam/metal/resin/halon
+
+/obj/effect/particle_effect/fluid/foam/metal/resin/halon/atmos_expose(datum/gas_mixture/air, exposed_temperature)
+	return // Doesn't dissolve in heat.
```
should override should_atmos_process and also maybe remove the element, otherwise this will process forever for no reason

Pickle-Coding replied:
Sure.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -820,28 +820,37 @@
 	if(new_heat_capacity > MINIMUM_HEAT_CAPACITY)
 		air.temperature = max(((temperature * old_heat_capacity - energy_used) / new_heat_capacity), TCMB)
+
+	var/turf/open/location = holder
```
are we guaranteed this?

Pickle-Coding replied:
I might test again if anything fucky happens in unusual gas mixtures, though I think it worked fine when I tested it 1-2 years ago.

LemonInTheDark replied:
I mean what happens if this reaction occurs in a canister, or a tank

Pickle-Coding replied:
I don't think it will pass the if check, but I'll double check tomorrow. It's typecasted as an open turf so I don't think doing this in a tank would make the location be anything.

Pickle-Coding replied:
My newer commit resolves this by the way.

LemonInTheDark commented on code/modules/atmospherics/gasmixtures/reactions.dm:
```diff
@@ -820,28 +820,37 @@
 		air.temperature = max(((temperature * old_heat_capacity - energy_used) / new_heat_capacity), TCMB)
+
+	var/turf/open/location = holder
+	var/obj/effect/particle_effect/fluid/foam/foam = locate() in location
+	var/obj/structure/foamedmetal/resin = locate() in location
+	if(heat_efficiency > HALON_COMBUSTION_MINIMUM_RESIN_MOLES && location && !foam && !resin)
+		var/datum/effect_system/fluid_spread/foam/metal/resin/halon/foaming = new
+		foaming.set_up(amount = HALON_COMBUSTION_RESIN_VOLUME, holder = holder, location = location)
+		foaming.start()
```
do we do similar checks elsewhere? would it be better to add a trait to the turf if it's foamed or not?

Pickle-Coding replied:
It probably would be cheaper and better to add and check for a trait for foam. Resin seems specific but I guess I could do that too.

LemonInTheDark replied:
concern is less efficency and more simplicity and the reduction of code debt, if we add a new thing that should block this I don't want to have to add a new locate you know?

Pickle-Coding replied:
That makes sense.

Pickle-Coding replied:
Looks like the only thing that checks for this is the resin mister on resin spray mode. Grenades and shit don't check for fluids. I guess that will be for another PR to fix this (maybe on foam initialize?).

LemonInTheDark commented on code/modules/atmospherics/machinery/components/electrolyzer/electrolyzer_reactions.dm:
```diff
@@ -89,30 +89,26 @@ GLOBAL_LIST_INIT(electrolyzer_reactions, electrolyzer_reactions_list())
-	air_mixture.gases[/datum/gas/nitrous_oxide][MOLES] -= reaction_efficency
-	air_mixture.gases[/datum/gas/halon][MOLES] += reaction_efficency
+	air_mixture.assert_gases(/datum/gas/bz, /datum/gas/oxygen, /datum/gas/halon)
+	var/reaction_efficency = min(air_mixture.gases[/datum/gas/bz][MOLES] * (1 - NUM_E ** (-0.5 * air_mixture.temperature * working_power / FIRE_MINIMUM_TEMPERATURE_TO_EXIST)), air_mixture.gases[/datum/gas/bz][MOLES])
```
what is goin on here

Pickle-Coding replied:
Faster the higher the temperature and electrolysis is, though it's decently fast in normal conditions anyways.

LemonInTheDark reviewed (changes requested):
Conceptually fine with this, unsure if it'll see use just cause gas can be a bit clunky and fires may be over before there's a point to it, but I'm game to give it a try

Pickle-Coding committed:
Fixes and shit.

Remove the atmos_sensitive element and override should_atmos_process, so it doesn't process for no reason.

Also fixes halon releasing resin while inside tanks and shit.

Pickle-Coding committed:
Compose the minimum mole requirements with defines.

Composes the minimum mole requirements for resin release with defines. This changes it slightly but it doesn't matter. Also adds a temperature scale define.

Pickle-Coding commented:
For some reason it bitches about when I use defines for the factor descriptions. Is there a way to get around that?

OverwatchVoice commented:
Glad to see an attempt at making halon more useful. I admit I tried making it just once, and after seeing two electrolysers produce a grand total of 9 mols,i quickly gave up on ever trying to use it.

As for the pluoxium, I don't think it will be pursued, as the tritium / co2 / o2 interaction exists, and can reliably be achieved just by using the burn chamber's output.

Ghommie commented on code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm:
```diff
@@ -416,6 +416,7 @@
 	alpha = 120
 	max_integrity = 10
 	pass_flags_self = PASSGLASS
+	var/list/ignored_gases = list(/datum/gas/nitrogen, /datum/gas/oxygen, /datum/gas/pluoxium, /datum/gas/halon)
```
```suggestion
	var/static/list/ignored_gases = typecacheof(list(
		/datum/gas/nitrogen,
		/datum/gas/oxygen,
		/datum/gas/pluoxium,
		/datum/gas/halon,
	))
```
Until subtypes that override their parent definition don't exist, it may as well be static.

Ghommie commented on code/game/objects/effects/effect_system/fluid_spread/effects_foam.dm:
```diff
@@ -432,11 +433,8 @@
-					continue
-				else
-					gases[gas_type][MOLES] = 0
+			if(!(gas_type in ignored_gases))
```
```suggestion
			if(!(ignored_gases[gas_type]))
```

Ghommie committed:
Apply suggestions from code review
